arangod/ClusterEngine/ClusterSelectivityEstimates.cpp

```diff
@@ -23,90 +23,80 @@
 
 #include "ClusterSelectivityEstimates.h"
 
-#include "Basics/ReadLocker.h"
-#include "Basics/WriteLocker.h"
 #include "Cluster/ClusterMethods.h"
 #include "Indexes/Index.h"
 #include "VocBase/LogicalCollection.h"
 
 using namespace arangodb;
-
+      
 ClusterSelectivityEstimates::ClusterSelectivityEstimates(LogicalCollection& collection)
-    : _collection(collection), _expireStamp(0.0) {}
+    : _collection(collection),
+      _updating(ATOMIC_FLAG_INIT) {}
 
 void ClusterSelectivityEstimates::flush() {
-  WRITE_LOCKER(lock, _lock);
-  _estimates.clear();
-  _expireStamp = 0.0;
-}
-
-IndexEstMap ClusterSelectivityEstimates::get(bool allowUpdate, TRI_voc_tid_t tid) const {
-  double now;
-
-  {
-    READ_LOCKER(readLock, _lock);
-
-    if (!allowUpdate) {
-      // return whatever is there. may be empty as well
-      return _estimates;
-    }
-
-    now = TRI_microtime();
-    if (!_estimates.empty() && _expireStamp > now) {
-      // already have an estimate, and it is not yet expired
-      return _estimates;
-    }
+  while (_updating.test_and_set()) {
+    std::this_thread::yield();
```

Comment:
Unfortunately, `atomic_flag` does not provide a load function, so I would rather keep `_updating` an `atomic<bool>` and only attempt the `exchange` if it is currently not set to true to avoid cache thrashing:
```
while (_updating.load(std::memory_order_relaxed) || _updating.exchange(true, std::memory_order_acquire)) {
  std::this_thread::yield();
}
```

Reply:
ok Test and test-and-set is probably better

Reply:
@mpoeter: right :smile: